add grpc broker.

morvencao opened the pull request:
This PR introduces a gRPC message broker, enabling the Maestro agent to subscribe to resource specifications and publish resource statuses, similar to how an MQTT broker operates.

depends on: https://github.com/open-cluster-management-io/sdk-go/pull/68

morvencao commented:
/assign @qiujian16 @clyang82 @skeeey

clyang82 commented:
/ok-to-test

clyang82 commented on templates/service-template.yml:
```diff
@@ -109,6 +109,16 @@ parameters:
+- name: GRPC_BROKER_BINDPORT
+  displayName: gRPC Broker Bindport
+  description: gRPC broker bind port
+  value: "8090"
```
`8091`?

morvencao replied:
good catch, updated this to `8091`.

clyang82 commented on templates/service-template.yml:
```diff
@@ -395,19 +406,37 @@ objects:
       ports:
-        - port: 8090
+        - name: grpc
+          port: 8090
```
Can we use ${GRPC_SERVER_BINDPORT}?

morvencao replied:
After changing to `${GRPC_SERVER_BINDPORT}`, I encountered the following issue:
```
json: cannot unmarshal string into Go struct field ServicePort.spec.ports.port of type int32
```

clyang82 commented on pkg/config/grpc_server.go:
```diff
@@ -28,7 +30,9 @@ func NewGRPCServerConfig() *GRPCServerConfig {
 	fs.BoolVar(&s.EnableGRPCServer, "enable-grpc-server", false, "Enable gRPC server")
-	fs.StringVar(&s.BindPort, "grpc-server-bindport", "8090", "gPRC server bind port")
+	fs.StringVar(&s.ServerBindPort, "grpc-server-bindport", "8090", "gPRC server bind port")
+	fs.BoolVar(&s.EnableGRPCBroker, "enable-grpc-broker", false, "Enable gRPC broker")
```
Can we use https://github.com/openshift-online/maestro/blob/b65178176cf59907febff9ac4388341eb0e7b801/pkg/config/message_broker.go#L31 to add `grpc` instead?

morvencao replied:
I'm still considering if this makes sense.
Currently, the message broker type is used for the Maestro server to publish resources and subscribe to resource status, while the gRPC broker is mainly for the agent to publish status and subscribe to resource spec.

Since the Maestro server and gRPC broker run together in the same process, the Maestro server doesn't use gRPC to publish resources or subscribe to resource status. Therefore, enabling the gRPC broker for the Maestro server doesn't change the message type; it merely forwards resources to gRPC broker subscribers (work agents).

So, I'm unsure if we should change the meaning of the message type to accommodate this case.

skeeey commented on cmd/maestro/servecmd/cmd.go:
```diff
@@ -75,6 +80,10 @@ func runServer(cmd *cobra.Command, args []string) {
 	go pulseServer.Start(ctx)
 	go controllersServer.Start(ctx)
+	// Start the GRPC broker if enabled
+	if grpcBroker != nil {
```
nit: may 
```
if environments.Environment().Config.GRPCServer.EnableGRPCBroker {
grpcBroker = server.NewGRPCBroker()
  go grpcBroker.Start(ctx)
}
```

skeeey commented on cmd/maestro/server/grpc_broker.go:
```diff
@@ -0,0 +1,532 @@
+		return nil, fmt.Errorf("unmatched consumer name %s for resource %s", consumerName, resourceID)
+	}
+
+	specEvent, err := api.JSONMAPToCloudEvent(found.Payload)
```
is the following part same with RESTful part? can we make a common function to handle this?

morvencao replied:
Yes, I created a common function (handleStatusUpdate) so we can handle the status event from either the MQTT or gRPC broker using the same logic.

skeeey commented on cmd/maestro/server/grpc_broker.go:
```diff
@@ -0,0 +1,532 @@
+	defer bkr.mu.Unlock()
+
+	close(bkr.subscribers[id].errChan)
+	delete(bkr.subscribers, id)
```
need a log like the register

morvencao replied:
added a log for unregister.

skeeey commented on pkg/db/event_advisory_lock.go:
```diff
@@ -0,0 +1,192 @@
+
+// NewEventAdvisoryLockFactory returns a new EventAdvisoryLockFactory with an AdvisoryLockStore
+// and a function to check if the consumer relevant to the current event is subscribed to the current gRPC broker.
+func NewEventAdvisoryLockFactory(connection SessionFactory, isConsumerSubscribed IsConsumerSubscribed) *EventAdvisoryLockFactory {
```
can we use a new controller instead of this? I think for grpc broker, we don't need a lock for spec event, when a spec event is send, the controller broadcast the event for every grpc broker, each broker only handle event with its consumers? for service status update request, we can keep the current logic

morvencao replied:
Last week, we discussed this with @qiujian16 and @clyang82. I understand that adding a new controller (specController without lock) means each instance must handle the event and add an <event, instance> record after the instance processes the event, similar to what we did for statusController. I'm still open to this and we can discuss it further in the future.

morvencao commented:
removed the multiple maestro server (grpc broker) instances for now in https://github.com/openshift-online/maestro/pull/157/commits/1083ec5ae48335075b8ea1868dd384722bc98ebb and added a TODO item for that, because it's not the required function for now and the implementation may impact other components.
/cc @qiujian16

skeeey commented on cmd/maestro/environments/framework.go:
```diff
@@ -173,26 +173,28 @@ func (e *Env) LoadClients() error {
+				glog.Errorf("Unable to build cloudevent source options: %s", err.Error())
 				return err
 			}
+			if cloudEventsSourceOptions != nil {
```
the cloudEventsSourceOptions will be nil under what cases?

morvencao replied:
looks like we don't need this check.

morvencao replied:
removed this check.

skeeey commented on cmd/maestro/servecmd/cmd.go:
```diff
@@ -38,12 +38,20 @@ func runServer(cmd *cobra.Command, args []string) {
+	// For gRPC, create a gRPC broker to handle resource spec and status events.
+	// For MQTT, create a Pulse server to handle resource spec and status events.
+	var eventServer server.EventServer
+	if environments.Environment().Config.MessageBroker.MessageBrokerType == "grpc" {
```
may add log for this part

morvencao replied:
added log.

skeeey commented:
LGTM

clyang82 reviewed (approved):
LGTM
